src/hotspot/share/oops/instanceKlass.cpp

```diff
@@ -2549,7 +2549,13 @@ void InstanceKlass::remove_unshareable_info() {
   _oop_map_cache = NULL;
   // clear _nest_host to ensure re-load at runtime
   _nest_host = NULL;
+#if !INCLUDE_CDS_JAVA_HEAP
   _package_entry = NULL;
+#else
+  if (!MetaspaceShared::use_full_module_graph()) {
+    _package_entry = NULL;
+  }
+#endif
```

Comment:
I think the `#if` part is not needed. `MetaspaceShared::use_full_module_graph()` returns false if `INCLUDE_CDS_JAVA_HEAP` is false.

Instead of `ArchiveBuilder::update_package_entry()`, maybe we can set `_package_entry` here?

```
if (!MetaspaceShared::use_full_module_graph()) {
  _package_entry = NULL;
} else if (DynamicDumpSharedSpaces) {
  if (!Metaspace::is_in_shared_metaspace(_package_entry)) {
    _package_entry = NULL;  
  }
} else {
  _package_entry =  PackageEntry::get_archived_entry(entry);
}
ArchiveUtils::mark_pointer((address**)&_package_entry);
```

The `mark_pointer` call is necessary if relocation happens during dumping.

For classes loaded by custom loaders, we don't archive their modules so get_archived_entry() should return a NULL package. I think for sanity, we should add an assert like:

```
if (ik->is_shared_unregistered_class()) {
  assert(_package_entry == NULL, "package entry for unregistered classes are not archived");
}
```

Reply:
Also need to assert `_package_entry == NULL` if the class is in an unnamed module.

Reply:
Hi Ioi,

Thanks for your review.
As discussed off-line, instead of adding assert, the `else` block of the code will be as follows:
```
    if (is_shared_unregistered_class()) {
      _package_entry = NULL;
    } else {
      _package_entry = PackageEntry::get_archived_entry(_package_entry);
    }
```
I've made other changes based on your comments. Please see updated webrev 01.

thanks,
Calvin